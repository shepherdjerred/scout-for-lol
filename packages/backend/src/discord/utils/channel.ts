import type { Channel, Message, MessageCreateOptions, MessagePayload } from "discord.js";

/**
 * Minimal interface for text channels that support sending messages
 */
export type SendableChannel = {
  send(content: string | MessagePayload | MessageCreateOptions): Promise<Message>;
};

/**
 * Check if a channel is text-based and return it with proper typing
 *
 * This is one of the rare cases where we use a type assertion, as Discord.js
 * channel types are complex and after checking isTextBased(), we know it has send()
 *
 * @param channel Channel to check
 * @returns SendableChannel if channel is text-based, undefined otherwise
 */
export function asTextChannel(channel: Channel): SendableChannel | undefined {
  if (!channel.isTextBased()) {
    return undefined;
  }

  // Type assertion needed: Discord.js's isTextBased() runtime check guarantees send() method exists
  // but TypeScript's type system can't express this relationship
  // eslint-disable-next-line no-restricted-syntax -- Discord.js runtime check guarantees type safety
  return channel as unknown as SendableChannel;
}
