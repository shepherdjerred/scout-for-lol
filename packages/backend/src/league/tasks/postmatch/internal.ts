import { MatchV5DTOs } from "twisted/dist/models-dto/index.js";
import { z } from "zod";
import { api } from "../../api/api.ts";
import {
  AttachmentBuilder,
  EmbedBuilder,
  Message,
  MessageCreateOptions,
  MessagePayload,
} from "discord.js";
import { matchToImage } from "@scout-for-lol/report";
import {
  ApplicationState,
  CompletedMatch,
  type DiscordChannelId,
  getLaneOpponent,
  invertTeam,
  type LeaguePuuid,
  LoadingScreenState,
  parseQueueType,
  parseTeam,
  Player,
  PlayerConfigEntry,
  type Rank,
} from "@scout-for-lol/data";
import { getState, setState } from "../../model/state.ts";
import { differenceWith, filter, map, pipe } from "remeda";
import { getOutcome } from "../../model/match.ts";
import { regionToRegionGroup } from "twisted/dist/constants/regions.js";
import { mapRegionToEnum } from "../../model/region.ts";
import { participantToChampion } from "../../model/champion.ts";
import { saveMatchToS3 } from "../../../storage/s3.ts";

export async function checkMatch(game: LoadingScreenState) {
  console.log(
    `[checkMatch] 🔍 Starting match check for matchId: ${game.matchId.toString()}`
  );
  console.log(`[checkMatch] 📊 Game details:`, {
    matchId: game.matchId.toString(),
    playersCount: game.players.length,
    queue: game.queue,
    added: game.added.toISOString(),
    uuid: game.uuid,
  });

  try {
    const firstPlayer = game.players[0];
    if (!firstPlayer) {
      throw new Error("No players found in game");
    }

    const region = mapRegionToEnum(
      firstPlayer.player.league.leagueAccount.region
    );
    console.log(`[checkMatch] 🌍 Mapped region: ${region}`);

    const regionGroup = regionToRegionGroup(region);
    console.log(`[checkMatch] 🌐 Region group: ${regionGroup}`);

    const matchIdForApi = `${region}_${game.matchId.toString()}`;
    console.log(`[checkMatch] 🔗 Calling API with matchId: ${matchIdForApi}`);

    const apiStartTime = Date.now();
    const response = await api.MatchV5.get(matchIdForApi, regionGroup);
    const apiTime = Date.now() - apiStartTime;

    console.log(
      `[checkMatch] ✅ API response received for match: ${game.matchId.toString()} in ${apiTime.toString()}ms`
    );
    console.log(
      `[checkMatch] 📊 Match status: ${response.response.info.gameEndTimestamp ? "FINISHED" : "IN_PROGRESS"}`
    );

    if (response.response.info.gameEndTimestamp) {
      console.log(
        `[checkMatch] 🏁 Match finished at: ${new Date(response.response.info.gameEndTimestamp).toISOString()}`
      );
      console.log(
        `[checkMatch] ⏱️  Game duration: ${response.response.info.gameDuration.toString()}s`
      );
    }

    return response.response;
  } catch (e) {
    console.error(
      `[checkMatch] ❌ Error occurred for match ${game.matchId.toString()}:`,
      e
    );
    const result = z.object({ status: z.number() }).safeParse(e);
    if (result.success) {
      console.log(
        `[checkMatch] 🔢 HTTP status code: ${result.data.status.toString()}`
      );
      if (result.data.status === 404) {
        // game not done
        console.log(
          `[checkMatch] ⏳ Match ${game.matchId.toString()} not finished yet (404)`
        );
        return undefined;
      }
      if (result.data.status === 403) {
        // Not recoverable: log and remove from queue
        console.error(
          `[checkMatch] 🚫 403 Forbidden for match ${game.matchId.toString()}, removing from queue`
        );
        const currentState = getState();
        const newGamesStarted = currentState.gamesStarted.filter(
          (g) => g.matchId !== game.matchId
        );
        setState({
          ...currentState,
          gamesStarted: newGamesStarted,
        });
        console.log(
          `[checkMatch] 🗑️  Removed match ${game.matchId.toString()} from tracking`
        );
        return undefined;
      }
    }
    console.error(
      `[checkMatch] 💥 Unhandled error for match ${game.matchId.toString()}:`,
      e
    );
    return undefined;
  }
}

export async function saveMatch(match: MatchV5DTOs.MatchDto): Promise<void> {
  console.log(`[saveMatch] 💾 Saving match: ${match.metadata.matchId}`);
  console.log(`[saveMatch] 📊 Match details:`, {
    matchId: match.metadata.matchId,
    participants: match.info.participants.length,
    gameMode: match.info.gameMode,
    queueId: match.info.queueId,
    gameEndTimestamp: match.info.gameEndTimestamp,
    gameDuration: match.info.gameDuration,
  });

  try {
    // Save the match to S3
    await saveMatchToS3(match);

    console.log(`[saveMatch] ✅ Successfully saved match: ${match.metadata.matchId}`);
  } catch (error) {
    console.error(`[saveMatch] ❌ Error saving match ${match.metadata.matchId}:`, error);

    // Don't throw the error to prevent disrupting the entire post-match flow
    // The match processing should continue even if S3 storage fails
    console.warn(`[saveMatch] ⚠️  Continuing post-match processing despite storage failure`);
  }
}

async function getImage(
  match: CompletedMatch
): Promise<[AttachmentBuilder, EmbedBuilder]> {
  console.log(`[getImage] 🖼️  Starting image generation for match`);
  console.log(`[getImage] 📊 Match details:`, {
    queueType: match.queueType,
    playersCount: match.players.length,
    durationInSeconds: match.durationInSeconds,
    teamsBlue: match.teams.blue.length,
    teamsRed: match.teams.red.length,
  });

  try {
    const imageStartTime = Date.now();
    const image = await matchToImage(match);
    const imageTime = Date.now() - imageStartTime;

    console.log(
      `[getImage] ✅ Image generated successfully in ${imageTime.toString()}ms, size: ${image.length.toString()} bytes`
    );

    const attachment = new AttachmentBuilder(image).setName("match.png");
    if (!attachment.name) {
      console.error(`[getImage] ❌ Attachment name is null`);
      throw new Error("[getImage] Attachment name is null");
    }
    console.log(
      `[getImage] 📎 Attachment created with name: ${attachment.name}`
    );

    const embed = {
      image: {
        url: `attachment://${attachment.name}`,
      },
    };
    console.log(`[getImage] 🎨 Embed created successfully`);

    return [attachment, new EmbedBuilder(embed)];
  } catch (error) {
    console.error(`[getImage] 💥 Error generating image:`, error);
    throw error;
  }
}

async function createMatchObj(
  state: LoadingScreenState,
  match: MatchV5DTOs.MatchDto,
  getPlayerFn: (playerConfig: PlayerConfigEntry) => Promise<Player>
) {
  console.log(
    `[createMatchObj] 🏗️  Starting match object creation for matchId: ${match.metadata.matchId}`
  );
  console.log(`[createMatchObj] 📊 State details:`, {
    playersCount: state.players.length,
    queue: state.queue,
    matchId: state.matchId.toString(),
    uuid: state.uuid,
  });

  // Get teams using backend/model/match.ts helpers
  const getTeams = (participants: MatchV5DTOs.ParticipantDto[]) => {
    console.log(
      `[createMatchObj] 👥 Processing ${participants.length.toString()} participants into teams`
    );
    const teams = {
      blue: pipe(participants.slice(0, 5), map(participantToChampion)),
      red: pipe(participants.slice(5, 10), map(participantToChampion)),
    };
    console.log(
      `[createMatchObj] 🔵 Blue team: ${teams.blue.length.toString()} players`
    );
    console.log(
      `[createMatchObj] 🔴 Red team: ${teams.red.length.toString()} players`
    );
    return teams;
  };

  const teams = getTeams(match.info.participants);
  const queueType = parseQueueType(match.info.queueId);
  console.log(
    `[createMatchObj] 🎯 Queue type: ${queueType ?? "unknown"} (ID: ${match.info.queueId.toString()}), participants: ${match.info.participants.length.toString()}`
  );

  // Gather all relevant players
  console.log(
    `[createMatchObj] 👥 Processing ${state.players.length.toString()} players from loading screen`
  );
  const players = await Promise.all(
    state.players.map(async (playerState, index) => {
      console.log(
        `[createMatchObj] 🔍 Processing player ${(index + 1).toString()}/${state.players.length.toString()}: ${playerState.player.alias}`
      );
      console.log(`[createMatchObj] 📋 Player details:`, {
        alias: playerState.player.alias,
        puuid: playerState.player.league.leagueAccount.puuid,
        region: playerState.player.league.leagueAccount.region,
        queue: state.queue,
      });

      // Find the participant in the match by puuid
      const participant = match.info.participants.find(
        (p) => p.puuid === playerState.player.league.leagueAccount.puuid
      );
      if (!participant) {
        console.error(
          `[createMatchObj] ❌ Unable to find participant for player ${playerState.player.alias}:`,
          playerState.player
        );
        console.error(
          `[createMatchObj] 📋 Available participants:`,
          match.info.participants.map((p) => ({
            puuid: p.puuid,
            summonerName: p.summonerName,
          }))
        );
        throw new Error(
          `unable to find participant for player ${JSON.stringify(playerState)}, match: ${JSON.stringify(match)}`
        );
      }

      console.log(
        `[createMatchObj] ✅ Found participant for player ${(index + 1).toString()}: ${participant.summonerName} (Champion: ${participant.championId.toString()})`
      );
      console.log(`[createMatchObj] 📊 Participant stats:`, {
        championId: participant.championId,
        teamId: participant.teamId,
        kills: participant.kills,
        deaths: participant.deaths,
        assists: participant.assists,
        win: participant.win,
      });

      const fullPlayer = await getPlayerFn(playerState.player);
      console.log(
        `[createMatchObj] ✅ Retrieved full player data for player ${(index + 1).toString()}: ${playerState.player.alias}`
      );

      let rankBeforeMatch: Rank | undefined = undefined;
      let rankAfterMatch: Rank | undefined = undefined;
      if (state.queue === "solo" || state.queue === "flex") {
        rankBeforeMatch = playerState.rank;
        rankAfterMatch = fullPlayer.ranks[state.queue];
        console.log(
          `[createMatchObj] 📊 Rank data for player ${(index + 1).toString()}:`,
          {
            before: rankBeforeMatch?.tier,
            after: rankAfterMatch?.tier,
          }
        );
      }

      const champion = participantToChampion(participant);
      console.log(`[createMatchObj] 🏆 Champion info:`, {
        championName: champion.championName,
        lane: champion.lane,
      });

      const team = parseTeam(participant.teamId);
      if (!team) {
        console.error(
          `[createMatchObj] ❌ Could not determine team for participant:`,
          participant
        );
        throw new Error(
          `Could not determine team for participant: ${JSON.stringify(participant)}`
        );
      }
      const enemyTeam = invertTeam(team);
      console.log(
        `[createMatchObj] ⚔️  Player team: ${team}, enemy team: ${enemyTeam}`
      );

      const outcome = getOutcome(participant);
      console.log(
        `[createMatchObj] 🎯 Match outcome for ${playerState.player.alias}: ${outcome}`
      );

      const laneOpponent = getLaneOpponent(champion, teams[enemyTeam]);
      console.log(
        `[createMatchObj] 🥊 Lane opponent for ${playerState.player.alias}:`,
        laneOpponent
      );

      console.log(
        `[createMatchObj] ✅ Player ${(index + 1).toString()} processed:`,
        {
          champion: champion.championName,
          team: team,
          outcome: outcome,
        }
      );

      return {
        playerConfig: fullPlayer.config,
        rankBeforeMatch,
        rankAfterMatch,
        wins:
          state.queue === "solo" || state.queue === "flex"
            ? (fullPlayer.ranks[state.queue]?.wins ?? undefined)
            : undefined,
        losses:
          state.queue === "solo" || state.queue === "flex"
            ? (fullPlayer.ranks[state.queue]?.losses ?? undefined)
            : undefined,
        champion,
        outcome: outcome,
        team: team,
        lane: champion.lane,
        laneOpponent: laneOpponent,
      };
    })
  );

  console.log(
    `[createMatchObj] ✅ Match object created successfully for matchId: ${match.metadata.matchId}`
  );
  console.log(`[createMatchObj] 📊 Final match object:`, {
    queueType: queueType,
    playersCount: players.length,
    durationInSeconds: match.info.gameDuration,
    teamsBlue: teams.blue.length,
    teamsRed: teams.red.length,
  });

  return {
    queueType,
    players,
    durationInSeconds: match.info.gameDuration,
    teams,
  };
}

export async function checkPostMatchInternal(
  state: ApplicationState,
  saveFn: (match: MatchV5DTOs.MatchDto) => Promise<void>,
  checkFn: (
    game: LoadingScreenState
  ) => Promise<MatchV5DTOs.MatchDto | undefined>,
  sendFn: (
    message: string | MessagePayload | MessageCreateOptions,
    channelId: string
  ) => Promise<Message<true> | Message<false>>,
  getPlayerFn: (playerConfig: PlayerConfigEntry) => Promise<Player>,
  getSubscriptionsFn: (
    playerIds: LeaguePuuid[]
  ) => Promise<{ channel: DiscordChannelId }[]>
) {
  console.log("=== POST-MATCH CHECK START ===");
  console.log(
    `[checkPostMatchInternal] Found ${state.gamesStarted.length.toString()} games in progress`
  );

  if (state.gamesStarted.length === 0) {
    console.log(
      "[checkPostMatchInternal] No games to check, skipping post-match processing"
    );
    return;
  }

  // Log details about each game being checked
  state.gamesStarted.forEach((game, index) => {
    console.log(
      `[checkPostMatchInternal] Game ${(index + 1).toString()}: matchId=${game.matchId.toString()}, players=${game.players.length.toString()}, added=${game.added.toISOString()}`
    );
  });

  console.log("[checkPostMatchInternal] Checking match API for all games");
  const games = await Promise.all(state.gamesStarted.map(checkFn));

  console.log(
    `[checkPostMatchInternal] API check results: ${games.length.toString()} responses`
  );
  games.forEach((game, index) => {
    if (game) {
      console.log(
        `[checkPostMatchInternal] Game ${(index + 1).toString()}: FINISHED - ${game.metadata.matchId.toString()}`
      );
    } else {
      console.log(
        `[checkPostMatchInternal] Game ${(index + 1).toString()}: STILL IN PROGRESS or ERROR`
      );
    }
  });

  console.log("[checkPostMatchInternal] Filtering finished games");
  const finishedGames = pipe(
    state.gamesStarted,
    map((game) => {
      const firstPlayer = game.players[0];
      if (!firstPlayer) {
        throw new Error("No players found in game");
      }
      const region = mapRegionToEnum(
        firstPlayer.player.league.leagueAccount.region
      );
      const fullMatchId = `${region}_${game.matchId.toString()}`;
      return [
        game,
        games.find((g) => g?.metadata.matchId === fullMatchId),
      ] satisfies [LoadingScreenState, MatchV5DTOs.MatchDto | undefined];
    }),
    filter(([_game, match]) => match != undefined)
    // this case is required to get rid of the undefined type
  ) as unknown as [LoadingScreenState, MatchV5DTOs.MatchDto][];

  console.log(
    `[checkPostMatchInternal] Found ${finishedGames.length.toString()} finished games to process`
  );

  if (finishedGames.length === 0) {
    console.log(
      "[checkPostMatchInternal] No finished games to process, ending post-match check"
    );
    return;
  }

  // TODO: send duo queue message
  console.log(
    "[checkPostMatchInternal] Processing finished games and sending messages"
  );
  await Promise.all(
    map(finishedGames, async ([state, matchDto]) => {
      console.log(
        `[checkPostMatchInternal] Processing finished game: ${matchDto.metadata.matchId.toString()}`
      );

      try {
        console.log(
          `[checkPostMatchInternal] Saving match: ${matchDto.metadata.matchId}`
        );
        await saveFn(matchDto);
        console.log(
          `[checkPostMatchInternal] Successfully saved match: ${matchDto.metadata.matchId}`
        );

        console.log(
          `[checkPostMatchInternal] Creating match object for: ${matchDto.metadata.matchId}`
        );
        const matchObj = await createMatchObj(state, matchDto, getPlayerFn);
        console.log(
          `[checkPostMatchInternal] Successfully created match object for: ${matchDto.metadata.matchId}`
        );

        console.log(
          `[checkPostMatchInternal] Generating image for: ${matchDto.metadata.matchId}`
        );
        const [attachment, embed] = await getImage(matchObj);
        console.log(
          `[checkPostMatchInternal] Successfully generated image for: ${matchDto.metadata.matchId}`
        );

        // figure out what channels to send the message to
        // server, see if they have a player in the game
        console.log(
          `[checkPostMatchInternal] Getting subscriptions for match: ${matchDto.metadata.matchId}`
        );
        const firstPlayer = state.players[0];
        if (!firstPlayer) {
          throw new Error("No players found in game");
        }
        const servers = await getSubscriptionsFn([
          firstPlayer.player.league.leagueAccount.puuid,
        ]);
        console.log(
          `[checkPostMatchInternal] Found ${servers.length.toString()} subscribed channels for match: ${matchDto.metadata.matchId.toString()}`
        );

        if (servers.length === 0) {
          console.log(
            `[checkPostMatchInternal] No subscribed channels found for match: ${matchDto.metadata.matchId}, skipping message sending`
          );
        } else {
          console.log(
            `[checkPostMatchInternal] Sending messages to ${servers.length.toString()} channels for match: ${matchDto.metadata.matchId}`
          );
          const promises = servers.map((server) => {
            console.log(
              `[checkPostMatchInternal] Sending message to channel: ${server.channel} for match: ${matchDto.metadata.matchId}`
            );
            return sendFn(
              { embeds: [embed], files: [attachment] },
              server.channel
            );
          });
          await Promise.all(promises);
          console.log(
            `[checkPostMatchInternal] Successfully sent messages to all channels for match: ${matchDto.metadata.matchId}`
          );
        }

        console.log(
          `[checkPostMatchInternal] Calculating new state after processing match: ${matchDto.metadata.matchId}`
        );
        const newState = getState();
        const newMatches = differenceWith(
          newState.gamesStarted,
          map(finishedGames, (game) => game[0]),
          (left, right) => left.uuid === right.uuid
        );

        console.log(
          `[checkPostMatchInternal] Saving updated state after processing match: ${matchDto.metadata.matchId}`
        );
        setState({
          ...newState,
          gamesStarted: newMatches,
        });
        console.log(
          `[checkPostMatchInternal] Successfully updated state, removed ${finishedGames.length.toString()} finished games`
        );
      } catch (error) {
        console.error(
          `[checkPostMatchInternal] Error processing match ${matchDto.metadata.matchId}:`,
          error
        );
        throw error; // Re-throw so it gets caught by logErrors
      }
    })
  );

  console.log("=== POST-MATCH CHECK COMPLETE ===");
}
