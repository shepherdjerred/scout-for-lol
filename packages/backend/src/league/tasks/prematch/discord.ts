import { getChampionName } from "twisted/dist/constants/champions.js";
import { CurrentGameInfoDTO } from "twisted/dist/models-dto/index.js";
import { findParticipant } from "../../api/index.ts";
import { PlayerConfigEntry } from "@scout/data";
import { QueueType } from "@scout/data";
import { map } from "remeda";

export function createDiscordMessage(
  players: PlayerConfigEntry[],
  game: CurrentGameInfoDTO,
  queueType: QueueType | undefined,
): string {
  const participants = players.map((player) => {
    const participant = findParticipant(player, game.participants);
    if (participant === undefined) {
      throw new Error(
        `unable to find participants: ${
          JSON.stringify(
            participants,
          )
        }, ${JSON.stringify(game)}`,
      );
    }
    return { player, participant };
  });

  // TODO: call API to get proper champion name
  const messages = map(participants, (participant) => {
    // TODO: can be removed after https://github.com/Sansossio/twisted/issues/128
    const championName = participant.participant.championId === 893
      ? "Aurora"
      : getChampionName(participant.participant.championId);
    return `${participant.player.name} (${
      championName.replaceAll("_", " ").toLowerCase().replace(/\b\w/g, (char) =>
        char.toUpperCase())
    })`;
  });

  let messageString = messages.join(", ");
  if (messages.length > 1) {
    const lastCommaIndex = messageString.lastIndexOf(",");
    messageString = `${messageString.substring(0, lastCommaIndex)}, and${
      messageString.substring(lastCommaIndex + 1)
    }`;
  }

  return `${messageString} started a ${queueType} game`;
}
