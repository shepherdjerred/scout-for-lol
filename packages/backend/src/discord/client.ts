import configuration from "../configuration";
import { Client, GatewayIntentBits } from "discord.js";
import { handleCommands } from "./commands/index";

const client = new Client({
  intents: [GatewayIntentBits.Guilds],
});

console.log("Logging into Discord");
await client.login(configuration.discordToken);
console.log("Logged into Discord");

client.on("ready", () => {
  handleCommands(client);
  console.log("Ready to handle Discord commands");
});

export default client;
