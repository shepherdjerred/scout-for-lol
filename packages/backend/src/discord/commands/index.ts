import { type Client, MessageFlags } from "discord.js";
import { executeSubscribe } from "./subscribe";
import { executeUnsubscribe } from "./unsubscribe";
import { executeListSubscriptions } from "./listSubscriptions";

export function handleCommands(client: Client) {
  client.on("interactionCreate", (interaction) => {
    void (async () => {
      if (!interaction.isChatInputCommand()) {
        return;
      }
      console.log(interaction);

      try {
        if (interaction.commandName === "subscribe") {
          await executeSubscribe(interaction);
        } else if (interaction.commandName === "unsubscribe") {
          await executeUnsubscribe(interaction);
        } else if (interaction.commandName === "listsubscriptions") {
          await executeListSubscriptions(interaction);
        } else {
          await interaction.reply("Unknown command");
        }
      } catch (error) {
        console.error(error);
        if (interaction.replied || interaction.deferred) {
          await interaction.followUp({
            content: "There was an error while executing this command!",
            flags: MessageFlags.Ephemeral,
          });
        } else {
          await interaction.reply({
            content: "There was an error while executing this command!",
            flags: MessageFlags.Ephemeral,
          });
        }
      }
    })();
  });
}
