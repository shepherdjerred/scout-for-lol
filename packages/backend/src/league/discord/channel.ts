import { type Message, MessageCreateOptions, MessagePayload } from "discord.js";

export function send(
  options: string | MessagePayload | MessageCreateOptions,
): Promise<Message<true> | Message<false>> {
  // TODO
  // const channel = await client.channels.fetch(
  //   configuration.leagueChannelId,
  // ) as TextChannel;
  // if (!channel) {
  //   throw new Error("invalid channel");
  // }
  // return channel.send(options);
  console.log(`Sending message: ${options}`);
  const promise = new Promise<Message<true> | Message<false>>(
    (resolve, _reject) => {
      resolve({} as Message<true>);
    },
  );
  return promise;
}
