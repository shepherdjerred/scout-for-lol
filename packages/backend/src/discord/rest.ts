import { REST, Routes } from "discord.js";
import { z } from "zod";
import { subscribeCommand } from "./commands/subscribe";
import { unsubscribeCommand } from "./commands/unsubscribe";
import configuration from "../configuration";
import { listSubscriptionsCommand } from "./commands/list-subscriptions";
import { debugCommand } from "./commands/debug";
import { competitionCommand } from "./commands/competition/index.js";
import { adminCommand } from "./commands/admin/index.js";
import { serverInfoCommand } from "./commands/server-info.js";

console.log("üîÑ Preparing Discord slash commands for registration");

const commands = [
  subscribeCommand.toJSON(),
  unsubscribeCommand.toJSON(),
  listSubscriptionsCommand.toJSON(),
  debugCommand.toJSON(),
  competitionCommand.toJSON(),
  adminCommand.toJSON(),
  serverInfoCommand.toJSON(),
];

console.log("üìã Commands to register:");
commands.forEach((command, index) => {
  console.log(`  ${(index + 1).toString()}. ${command.name}: ${command.description}`);
});

console.log("üîë Initializing Discord REST client");
const rest = new REST().setToken(configuration.discordToken);

void (async () => {
  try {
    console.log(`üöÄ Starting registration of ${commands.length.toString()} application (/) commands`);
    console.log(`üéØ Target application ID: ${configuration.applicationId}`);

    const startTime = Date.now();
    const data = await rest.put(Routes.applicationCommands(configuration.applicationId), { body: commands });
    const registrationTime = Date.now() - startTime;

    console.log(
      `‚úÖ Successfully registered ${commands.length.toString()} application (/) commands in ${registrationTime.toString()}ms`,
    );

    // Log details about registered commands
    const CommandSchema = z.object({ name: z.string(), id: z.string() });
    const commandsResult = z.array(CommandSchema).safeParse(data);
    if (commandsResult.success) {
      console.log("üìù Registered commands details:");
      commandsResult.data.forEach((command, index) => {
        console.log(`  ${(index + 1).toString()}. ${command.name} (ID: ${command.id})`);
      });
    }

    console.log("üéâ Discord command registration completed successfully");
  } catch (error) {
    console.error("‚ùå Failed to register Discord commands:", error);

    // Log additional error context
    const ErrorDetailsSchema = z.object({ name: z.string(), message: z.string(), stack: z.string().optional() });
    const errorResult = ErrorDetailsSchema.safeParse(error);
    if (errorResult.success) {
      console.error("‚ùå Error name:", errorResult.data.name);
      console.error("‚ùå Error message:", errorResult.data.message);
      if (errorResult.data.stack) {
        console.error("‚ùå Error stack:", errorResult.data.stack);
      }
    }

    // Check for specific Discord API errors
    const objectResult = z.object({ status: z.unknown() }).catchall(z.unknown()).safeParse(error);
    if (objectResult.success) {
      const discordError = objectResult.data;
      console.error("‚ùå HTTP Status:", discordError.status);
      console.error("‚ùå Response body:", discordError["rawError"] ?? discordError["body"]);
    }

    process.exit(1);
  }
})();
