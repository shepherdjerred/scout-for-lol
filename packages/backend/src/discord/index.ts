import "./rest.ts";
import "./client.ts";
