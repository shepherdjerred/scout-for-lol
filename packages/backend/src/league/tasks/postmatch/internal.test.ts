import { MatchV5DTOs } from "twisted/dist/models-dto/index.js";
import {
  ApplicationState,
  Player,
  PlayerConfigEntry,
} from "@scout-for-lol/data";
import { send } from "../../discord/channel";
import { checkPostMatchInternal } from "./internal";
import { test, expect } from "bun:test";
import { Message, MessageCreateOptions, MessagePayload } from "discord.js";
import {
  DiscordAccountIdSchema,
  DiscordChannelIdSchema,
  LeaguePuuidSchema,
} from "@scout-for-lol/data";

const testdataPath = new URL("testdata/match.json", import.meta.url);

test("postmatch", async () => {
  const state: ApplicationState = {
    gamesStarted: [
      {
        uuid: "uuid",
        added: new Date(),
        matchId: 1,
        players: [
          {
            player: {
              alias: "name",
              league: {
                leagueAccount: {
                  puuid: LeaguePuuidSchema.parse(
                    "XtEsV464OFaO3c0_q9REa6wYF0HpC2LK4laLnyM7WhfAVeuDz9biieJ5ZRD049AUCBjLjyBeeezTaw",
                  ),
                  region: "AMERICA_NORTH",
                },
              },
              discordAccount: {
                id: DiscordAccountIdSchema.parse("123456789012345678"),
              },
            },
            rank: { division: 3, tier: "gold", lp: 11, wins: 10, losses: 20 },
          },
        ],
      },
    ],
  };
  const saveMatchFn = async (_: MatchV5DTOs.MatchDto) => {
    // do nothing
  };
  const sendFn = (async (
    message: string | MessagePayload | MessageCreateOptions,
  ): Promise<Message<true> | Message<false>> => {
    expect(message).toMatchSnapshot();
    return Promise.resolve({} as Message<true> | Message<false>);
  }) satisfies typeof send;
  const checkMatchFn = async () => {
    const exampleMatch = JSON.parse(
      await Bun.file(testdataPath).text(),
    ) as MatchV5DTOs.MatchDto;
    return exampleMatch;
  };
  const getPlayerFn = (_: PlayerConfigEntry): Promise<Player> => {
    return Promise.resolve({
      config: {
        alias: "name",
        league: {
          leagueAccount: {
            puuid: LeaguePuuidSchema.parse(
              "XtEsV464OFaO3c0_q9REa6wYF0HpC2LK4laLnyM7WhfAVeuDz9biieJ5ZRD049AUCBjLjyBeeezTaw",
            ),
            region: "AMERICA_NORTH",
          },
        },
        discordAccount: {
          id: DiscordAccountIdSchema.parse("12345678901234567"),
        },
      },
      ranks: {
        solo: { division: 3, tier: "gold", lp: 11, wins: 10, losses: 20 },
      },
    } satisfies Player);
  };
  const getSubscriptionsFn = () => {
    return Promise.resolve([
      { channel: DiscordChannelIdSchema.parse("12345678901234567") },
    ]);
  };

  await checkPostMatchInternal(
    state,
    saveMatchFn,
    checkMatchFn,
    sendFn,
    getPlayerFn,
    getSubscriptionsFn,
  );
});
