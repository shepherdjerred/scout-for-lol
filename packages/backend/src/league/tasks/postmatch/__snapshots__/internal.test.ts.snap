export const snapshot = {};

snapshot[`postmatch 1`] = `
{
  embeds: [
    EmbedBuilder {
      data: {
        image: {
          url: "attachment://match.png",
        },
      },
    },
  ],
  files: [
    AttachmentBuilder {
      attachment: <Buffer 89 50 4e 47 0d 0a 1a 0a 00 00 00 0d 49 48 44 52 00 00 12 98 00 00 0d ac 08 06 00 00 00 10 64 79 50 00 2f 5b 29 49 44 41 54 78 9c ec dc 7f 8c 95 f5 bd ... >,
      description: undefined,
      name: "match.png",
    },
  ],
}
`;
